core/src/main/java/org/apache/iceberg/ManifestsWriter.java

```diff
@@ -0,0 +1,115 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.collect.Lists;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.io.OutputFile;
+
+/**
+ * A manifest writer that splits entries into multiple files in order to produce files of the configured size.
+ */
+public class ManifestsWriter implements Closeable {
+  private final PartitionSpec spec;
+  private final Long snapshotId;
+  private final Supplier<OutputFile> outputFileSupplier;
+  private final long targetManifestSizeBytes;
+  private final List<ManifestFile> manifests;
+  private ManifestWriter writer;
+  private long currentNumEntries;
+
+  public ManifestsWriter(PartitionSpec spec, Long snapshotId,
+                         Supplier<OutputFile> outputFileSupplier,
+                         long targetManifestSizeBytes) {
+    this.spec = spec;
+    this.snapshotId = snapshotId;
+    this.outputFileSupplier = outputFileSupplier;
+    this.targetManifestSizeBytes = targetManifestSizeBytes;
+    this.manifests = Lists.newArrayList();
+    this.currentNumEntries = 0;
+  }
+
+  public ManifestsWriter(PartitionSpec spec,
+                         Supplier<OutputFile> outputFileSupplier,
+                         long targetManifestSizeBytes) {
+    this(spec, null, outputFileSupplier, targetManifestSizeBytes);
+  }
+
+  public void add(DataFile addedFile) {
+    lazyWriter().add(addedFile);
+    currentNumEntries++;
+  }
+
+  void add(ManifestEntry entry) {
+    lazyWriter().add(entry);
+    currentNumEntries++;
+  }
+
+  public void existing(DataFile existingFile, long fileSnapshotId) {
+    lazyWriter().existing(existingFile, fileSnapshotId);
+    currentNumEntries++;
+  }
+
+  void existing(ManifestEntry entry) {
+    lazyWriter().existing(entry);
+    currentNumEntries++;
+  }
+
+  public void delete(DataFile deletedFile) {
+    lazyWriter().delete(deletedFile);
+    currentNumEntries++;
+  }
+
+  void delete(ManifestEntry entry) {
+    lazyWriter().delete(entry);
+    currentNumEntries++;
+  }
+
+  public void close() {
+    if (writer != null) {
+      try {
+        writer.close();
+        manifests.add(writer.toManifestFile());
+        currentNumEntries = 0;
+      } catch (IOException e) {
+        throw new RuntimeIOException(e);
+      }
+    }
+  }
+
+  public List<ManifestFile> manifests() {
+    return manifests;
+  }
+
+  private ManifestWriter lazyWriter() {
+    // verify the size of the current manifest every 25 entries to avoid calling writer.length() every time
+    if (writer == null) {
+      writer = new ManifestWriter(spec, outputFileSupplier.get(), snapshotId);
+    } else if (currentNumEntries % 25 == 0 && writer.length() >= targetManifestSizeBytes) {
```

Comment:
Minor: I usually like to add a constant for the check interval so it can be configured easily.